**Table of contents**<a id='toc0_'></a>    
- [Importing Data](#toc1_)    
  - [Exercise 4.8 Practice](#toc1_1_)    
- [Deriving columns to Analyze Customers](#toc2_)    
  - [Creating a subset of aggregate order number mean per department id](#toc2_1_)    
  - [Analyzing Customers based on a loyalty flag](#toc2_2_)    
  - [Creating Customers Spending Flag](#toc2_3_)    
  - [Creating Order Frequency Flag](#toc2_4_)    
- [Exporting Changes as Pickle File](#toc3_)    

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

# <a id='toc1_'></a>[Importing Data](#toc0_)

In [1]:
#Importing Libraries
import numpy as np
import pandas as pd
import os

In [2]:
# Importing Subset of merged pkl file
Path = r'D:\Data Analysis\01-08-2025 Instacart Basket Analysis\Data'
ord_pro_merge = pd.read_pickle(os.path.join(Path, 'Prepared Data', 'ord_pro_busyderived.pkl'))

In [3]:
ord_pro_merge = ord_pro_merge.drop(columns=['eval_set'])
ord_pro_merge.tail()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
32404854  1285346  206208            48                  1                 11   
32404855  1285346  206208            48                  1                 11   
32404856  1285346  206208            48                  1                 11   
32404857  3186442  206209             6                  0                 16   
32404858  3186442  206209             6                  0                 16   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
32404854                     5.0       35939                  6          1   
32404855                     5.0       43086                  7          1   
32404856                     5.0       34282                  8          0   
32404857                     3.0       14197                  1          0   
32404858                     3.0        6187                  2          1   

                     product_name  aisle_id  department_id  prices  \
32404854  Organic Jalapeno Pepper        83              4    13.8   
32404855       Super Greens Salad       123              4     8.0   
32404856  Original Sandwich Bread        58              1    15.0   
32404857             Tomato Paste         9              9     5.6   
32404858       Raisin Bran Cereal       121             14     5.5   

         price_range    busiest_days busiest_period_of_day  
32404854   Mid Range  Second Busiest           Most Orders  
32404855   Mid Range  Second Busiest           Most Orders  
32404856   Mid Range  Second Busiest           Most Orders  
32404857   Mid Range     Busiest Day           Most Orders  
32404858   Mid Range     Busiest Day           Most Orders

## <a id='toc1_1_'></a>[Exercise 4.8 Practice](#toc0_)

In [6]:
df = ord_pro_merge[:1_000_000]

In [7]:
df.shape

(1000000, 17)

In [8]:
df.groupby('product_name')

In [9]:
df.groupby('department_id')['order_number'].mean()

department_id
1     15.279343
2     17.261884
3     17.151435
4     17.813307
5     15.425825
6     16.092425
7     17.574464
8     15.395840
9     16.039195
10    20.133903
11    16.346300
12    15.929525
13    16.712106
14    16.811298
15    16.119052
16    17.728944
17    16.048428
18    19.170604
19    17.291087
20    16.627686
21    22.139960
Name: order_number, dtype: float64

In [17]:
df.groupby('department_id').agg({'order_number':'mean'}).sort_values('order_number')

order_number
department_id              
1                 15.279343
8                 15.395840
5                 15.425825
12                15.929525
9                 16.039195
17                16.048428
6                 16.092425
15                16.119052
11                16.346300
20                16.627686
13                16.712106
14                16.811298
3                 17.151435
2                 17.261884
19                17.291087
7                 17.574464
16                17.728944
4                 17.813307
18                19.170604
10                20.133903
21                22.139960

In [19]:
df.groupby('department_id').agg({'order_number':['mean', 'max', 'min']})

order_number        
                      mean max min
department_id                     
1                15.279343  99   1
2                17.261884  99   1
3                17.151435  99   1
4                17.813307  99   1
5                15.425825  99   1
6                16.092425  99   1
7                17.574464  99   1
8                15.395840  99   1
9                16.039195  99   1
10               20.133903  99   1
11               16.346300  99   1
12               15.929525  98   1
13               16.712106  99   1
14               16.811298  99   1
15               16.119052  99   1
16               17.728944  99   1
17               16.048428  99   1
18               19.170604  99   1
19               17.291087  99   1
20               16.627686  99   1
21               22.139960  99   1

# <a id='toc2_'></a>[Deriving columns to Analyze Customers](#toc0_)

## <a id='toc2_1_'></a>[Creating a subset of aggregate order number mean per department id](#toc0_)

In [ ]:
ord_pro_merge.groupby('department_id')['order_number'].mean()

In [5]:
ord_pro_merge.groupby('department_id').agg({'order_number':['mean', 'median']})

order_number       
                      mean median
department_id                    
1                15.457838   10.0
2                17.277920   11.0
3                17.170395   11.0
4                17.811403   11.0
5                15.215751    9.0
6                16.439806   10.0
7                17.225802   11.0
8                15.340650   10.0
9                15.895474   10.0
10               20.197148   14.0
11               16.170638   10.0
12               15.887671   10.0
13               16.583536   10.0
14               16.773669   11.0
15               16.165037   10.0
16               17.665606   11.0
17               15.694469   10.0
18               19.310397   13.0
19               17.177343   11.0
20               16.473447   10.0
21               22.902379   16.0

3. The results show a slight decrease in the mean of most departments for the entire dataframe, when compared to the 1,000,000 row subset (except for department IDs 1,2,3,6,10,15,18,21), which could mean the subset is an over-representation of customer ordering behavior. This could also mean the dataframe is positively skewed with a few high data points included in the subset that increase the mean over the median.

## <a id='toc2_2_'></a>[Analyzing Customers based on a loyalty flag](#toc0_)

In [ ]:
# 4. Creating a max_order column using transform to base flagging

ord_pro_merge['max_order'] = ord_pro_merge.groupby('user_id')['order_number'].transform('max')

In [7]:
#Creating flag column designating loyalty level of customers based on max orders
ord_pro_merge.loc[ord_pro_merge['max_order']>40, 'loyalty_flag'] = 'Loyal Customer'
ord_pro_merge.loc[(ord_pro_merge['max_order']>10) & (ord_pro_merge['max_order']<= 40), 'loyalty_flag'] = 'Regular Customer'
ord_pro_merge.loc[ord_pro_merge['max_order']<=10, 'loyalty_flag'] = 'New Customer'
ord_pro_merge['loyalty_flag'].value_counts(dropna=False)

loyalty_flag
Regular Customer    15876776
Loyal Customer      10284093
New Customer         6243990
Name: count, dtype: int64

In [8]:
ord_pro_merge.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  3108588       1             8                  1                 14   
1  3108588       1             8                  1                 14   
2  3108588       1             8                  1                 14   
3  3108588       1             8                  1                 14   
4  3108588       1             8                  1                 14   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    14.0       12427                  1          1   
1                    14.0         196                  2          1   
2                    14.0       10258                  3          1   
3                    14.0       25133                  4          1   
4                    14.0       46149                  5          0   

            product_name  aisle_id  department_id  prices price_range  \
0    Original Beef Jerky        23             19     4.4   Low Range   
1                   Soda        77              7     9.0   Mid Range   
2             Pistachios       117             19     3.0   Low Range   
3  Organic String Cheese        21             16     8.6   Mid Range   
4      Zero Calorie Cola        77              7    13.4   Mid Range   

     busiest_days busiest_period_of_day  max_order  loyalty_flag  
0  Second Busiest           Most Orders         10  New Customer  
1  Second Busiest           Most Orders         10  New Customer  
2  Second Busiest           Most Orders         10  New Customer  
3  Second Busiest           Most Orders         10  New Customer  
4  Second Busiest           Most Orders         10  New Customer

In [9]:
# 5. Prices variation analysis among customers per loyalty flag
ord_pro_merge.groupby('loyalty_flag')['prices'].describe()

count       mean         std  min  25%  50%   75%  \
loyalty_flag                                                               
Loyal Customer    10284093.0  10.386338  327.981171  1.0  4.2  7.4  11.2   
New Customer       6243990.0  13.294668  597.521606  1.0  4.2  7.4  11.3   
Regular Customer  15876776.0  12.495728  539.679016  1.0  4.2  7.4  11.3   

                      max  
loyalty_flag               
Loyal Customer    99999.0  
New Customer      99999.0  
Regular Customer  99999.0

In [10]:
ord_pro_merge.groupby('loyalty_flag').agg({'prices':['mean', 'std']})

prices            
                       mean         std
loyalty_flag                           
Loyal Customer    10.386335  328.017792
New Customer      13.294669  597.560303
Regular Customer  12.495717  539.720947

The difference in above indicate it is better to use the agg() method for mean and standard deviation calculations than the describe and that precision may improve if the price data type is increased to int64. However, above results includes duplicate price values that without order numbers could obfuscate the analysis.

In [11]:
#creating a value count subset of prices to understand unique values
Frequency = ord_pro_merge.groupby('loyalty_flag')['prices'].value_counts().sort_index(level='prices', ascending=False)

In [12]:
#Converting subset to data frame to analyze
df_frequency = pd.DataFrame(Frequency).reset_index()
df_frequency.columns

Index(['loyalty_flag', 'prices', 'count'], dtype='object')

In [13]:
df_frequency.head(12)

loyalty_flag   prices  count
0   Regular Customer  99999.0    413
1       New Customer  99999.0    202
2     Loyal Customer  99999.0     83
3   Regular Customer  14900.0   2235
4       New Customer  14900.0    947
5     Loyal Customer  14900.0   1247
6   Regular Customer     25.0   7084
7       New Customer     25.0   2851
8     Loyal Customer     25.0   3837
9   Regular Customer     24.9   3165
10      New Customer     24.9   1284
11    Loyal Customer     24.9   1332

Creating an equivalent multiplier that reflects weights for price counts based on amount of records included in each loyalty flag to create a price equivalent count that can be easily compared.
Loyalty Flag	  Record Count	Equivalence multiplier
Regular Cust  	15876776	    1
Loyal Customer	10284093	    1.543818789
New Customer	  6243990	      2.542729248

In [14]:
# Creating an equivalent multiplier
def equiv_multiplier(row):
      if row['loyalty_flag'] == 'Loyal Customer':
        return row['count'] * 1.543818789
      elif row['loyalty_flag'] == 'New Customer':
        return row['count'] * 2.542729248
      else:
        return row['count']
df_frequency['equivalent'] = df_frequency.apply(equiv_multiplier, axis=1)
df_frequency.head(10)

loyalty_flag   prices  count   equivalent
0  Regular Customer  99999.0    413   413.000000
1      New Customer  99999.0    202   513.631308
2    Loyal Customer  99999.0     83   128.136959
3  Regular Customer  14900.0   2235  2235.000000
4      New Customer  14900.0    947  2407.964598
5    Loyal Customer  14900.0   1247  1925.142030
6  Regular Customer     25.0   7084  7084.000000
7      New Customer     25.0   2851  7249.321086
8    Loyal Customer     25.0   3837  5923.632693
9  Regular Customer     24.9   3165  3165.000000

In [15]:
#Used AI to obtain the below lambda function to calculate statistics for unique items only.
ord_pro_merge.groupby('loyalty_flag')['prices'].agg(
    count=lambda x: x.drop_duplicates().count(),
    mean=lambda x: x.drop_duplicates().mean(),
    median=lambda x: x.drop_duplicates().median(),
    std=lambda x: x.drop_duplicates().std(),
    Q1=lambda x: x.drop_duplicates().quantile(0.25),
    Q3=lambda x: x.drop_duplicates().quantile(0.75)
)

count        mean  median          std     Q1      Q3
loyalty_flag                                                           
Loyal Customer      241  489.595398   13.00  6507.717773  7.000  19.100
New Customer        242  487.647888   13.05  6494.272461  7.025  19.075
Regular Customer    242  487.647919   13.05  6494.272461  7.025  19.075

The above unique price statistics reflect contradictory information where new customers are no longer showing higher values, as compared to previous results or non-unique values. As can be seen, the higher unique price mean of $489.59 and lower unique price counts of 241 of loyal customers reveal a taste for higher priced products, however the higher standard deviation may also indicate they tend to buy a higher variability of products (with difference prices), which don't seem to be the most expensive products as shown in the df_frequency dataframe above. While new and regular customers seem to buy more items of similar prices that could mean they are interested in only certain types of products. 

## <a id='toc2_3_'></a>[Creating Customers Spending Flag](#toc0_)

In [16]:
# 6. Creating column to use in identify types of spenders
ord_pro_merge['mean_price'] = ord_pro_merge.groupby('user_id')['prices'].transform('mean')
ord_pro_merge.shape

(32404859, 19)

In [ ]:
ord_pro_merge.loc[ord_pro_merge['mean_price']<10, 'spender_type'] = 'Low Spender'
ord_pro_merge.loc[ord_pro_merge['mean_price']>= 10, 'spender_type'] = 'High Spender'
ord_pro_merge.tail()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
32404854  1285346  206208            48                  1                 11   
32404855  1285346  206208            48                  1                 11   
32404856  1285346  206208            48                  1                 11   
32404857  3186442  206209             6                  0                 16   
32404858  3186442  206209             6                  0                 16   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
32404854                     5.0       35939                  6          1   
32404855                     5.0       43086                  7          1   
32404856                     5.0       34282                  8          0   
32404857                     3.0       14197                  1          0   
32404858                     3.0        6187                  2          1   

                     product_name  aisle_id  department_id  prices  \
32404854  Organic Jalapeno Pepper        83              4    13.8   
32404855       Super Greens Salad       123              4     8.0   
32404856  Original Sandwich Bread        58              1    15.0   
32404857             Tomato Paste         9              9     5.6   
32404858       Raisin Bran Cereal       121             14     5.5   

         price_range    busiest_days busiest_period_of_day  max_order  \
32404854   Mid Range  Second Busiest           Most Orders         49   
32404855   Mid Range  Second Busiest           Most Orders         49   
32404856   Mid Range  Second Busiest           Most Orders         49   
32404857   Mid Range     Busiest Day           Most Orders         13   
32404858   Mid Range     Busiest Day           Most Orders         13   

              loyalty_flag  mean_price spender_type  
32404854    Loyal Customer    8.366618  Low Spender  
32404855    Loyal Customer    8.366618  Low Spender  
32404856    Loyal Customer    8.366618  Low Spender  
32404857  Regular Customer    7.058915  Low Spender  
32404858  Regular Customer    7.058915  Low Spender

## <a id='toc2_4_'></a>[Creating Order Frequency Flag](#toc0_)

In [ ]:
# 7. Creating customer frequency flag
order_median = ord_pro_merge.groupby('user_id')['days_since_prior_order'].median()
order_median.value_counts(dropna=False)

days_since_prior_order
30.0    34051
7.0     21586
8.0     13062
6.0     11100
14.0     9588
        ...  
27.5      125
2.5        72
1.5        28
0.5        21
NaN         1
Name: count, Length: 62, dtype: int64

In [19]:
Order_regularity = []

for dspo in order_median:
  if dspo > 20:
      Order_regularity.append('Non-frequent customer')
  elif dspo > 10 and dspo <= 20:
      Order_regularity.append('Regular Customer')
  else:
      Order_regularity.append('Frequent Customer')

In [ ]:
# Mapping orders frequency flag with help of AI
reg_mapping = dict(zip(order_median.index, Order_regularity))
ord_pro_merge['orders_frequency'] = ord_pro_merge['user_id'].map(reg_mapping)
ord_pro_merge.head()

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0  3108588       1             8                  1                 14   
1  3108588       1             8                  1                 14   
2  3108588       1             8                  1                 14   
3  3108588       1             8                  1                 14   
4  3108588       1             8                  1                 14   

   days_since_prior_order  product_id  add_to_cart_order  reordered  \
0                    14.0       12427                  1          1   
1                    14.0         196                  2          1   
2                    14.0       10258                  3          1   
3                    14.0       25133                  4          1   
4                    14.0       46149                  5          0   

            product_name  ...  department_id  prices  price_range  \
0    Original Beef Jerky  ...             19     4.4    Low Range   
1                   Soda  ...              7     9.0    Mid Range   
2             Pistachios  ...             19     3.0    Low Range   
3  Organic String Cheese  ...             16     8.6    Mid Range   
4      Zero Calorie Cola  ...              7    13.4    Mid Range   

     busiest_days busiest_period_of_day max_order  loyalty_flag mean_price  \
0  Second Busiest           Most Orders        10  New Customer   6.367797   
1  Second Busiest           Most Orders        10  New Customer   6.367797   
2  Second Busiest           Most Orders        10  New Customer   6.367797   
3  Second Busiest           Most Orders        10  New Customer   6.367797   
4  Second Busiest           Most Orders        10  New Customer   6.367797   

   spender_type       orders_frequency  
0   Low Spender  Non-frequent customer  
1   Low Spender  Non-frequent customer  
2   Low Spender  Non-frequent customer  
3   Low Spender  Non-frequent customer  
4   Low Spender  Non-frequent customer  

[5 rows x 21 columns]

In [ ]:
# Reviewing status of NAN values
ord_pro_merge[ord_pro_merge.isna().any(axis=1)]

order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
11        1374495       3             1                  1                 14   
12        1374495       3             1                  1                 14   
13        1374495       3             1                  1                 14   
14        1374495       3             1                  1                 14   
15        1374495       3             1                  1                 14   
...           ...     ...           ...                ...                ...   
32404591    28210  206201             1                  6                 14   
32404643  3189322  206206             1                  3                 18   
32404644  3189322  206206             1                  3                 18   
32404645  3189322  206206             1                  3                 18   
32404646  3189322  206206             1                  3                 18   

          days_since_prior_order  product_id  add_to_cart_order  reordered  \
11                           NaN        9387                  1          0   
12                           NaN       17668                  2          0   
13                           NaN       15143                  3          0   
14                           NaN       16797                  4          0   
15                           NaN       39190                  5          0   
...                          ...         ...                ...        ...   
32404591                     NaN       46607                 17          0   
32404643                     NaN       13817                  1          0   
32404644                     NaN       24099                  2          0   
32404645                     NaN       47011                  3          0   
32404646                     NaN       17734                  4          0   

                                               product_name  ...  \
11                                      Granny Smith Apples  ...   
12          Unsweetened Chocolate Almond Breeze Almond Milk  ...   
13                                           Blueberry Pint  ...   
14                                             Strawberries  ...   
15                          Vanilla Unsweetened Almond Milk  ...   
...                                                     ...  ...   
32404591                        StackerMallows Marshmallows  ...   
32404643  \"Im Pei-nut Butter\" Double Chocolate Cookie ...  ...   
32404644  Snickerdoodle Cookie with Salted Caramel Ice C...  ...   
32404645           Organic Ginger Limeade Fruit Juice Drink  ...   
32404646                  Gluten Free Soy Sauce Tamari Lite  ...   

          department_id  prices  price_range    busiest_days  \
11                    4     8.8    Mid Range  Second Busiest   
12                   16     6.6    Mid Range  Second Busiest   
13                    4     6.0    Mid Range  Second Busiest   
14                    4     1.3    Low Range  Second Busiest   
15                   16    12.3    Mid Range  Second Busiest   
...                 ...     ...          ...             ...   
32404591             13    12.6    Mid Range  Regularly Busy   
32404643              1    10.8    Mid Range  Second Slowest   
32404644              1    10.6    Mid Range  Second Slowest   
32404645              7    14.5    Mid Range  Second Slowest   
32404646              6     7.5    Mid Range  Second Slowest   

         busiest_period_of_day max_order      loyalty_flag mean_price  \
11                 Most Orders        12  Regular Customer   8.197727   
12                 Most Orders        12  Regular Customer   8.197727   
13                 Most Orders        12  Regular Customer   8.197727   
14                 Most Orders        12  Regular Customer   8.197727   
15                 Most Orders        12  Regular Customer   8.197727   
...                        ...       ...               ...        ...   
32

In [22]:
ord_pro_merge.info

<bound method DataFrame.info of          order_id user_id  order_number  order_day_of_week  order_hour_of_day  \
0         3108588       1             8                  1                 14   
1         3108588       1             8                  1                 14   
2         3108588       1             8                  1                 14   
3         3108588       1             8                  1                 14   
4         3108588       1             8                  1                 14   
...           ...     ...           ...                ...                ...   
32404854  1285346  206208            48                  1                 11   
32404855  1285346  206208            48                  1                 11   
32404856  1285346  206208            48                  1                 11   
32404857  3186442  206209             6                  0                 16   
32404858  3186442  206209             6                  0                 16

# <a id='toc3_'></a>[Exporting Changes as Pickle File](#toc0_)

In [24]:
ord_pro_merge.to_pickle(os.path.join(Path, 'Prepared Data', 'ord_pro_flagged.pkl'))